episodes/fig/source/06-blurring/create_blur_animation.py

```diff
@@ -0,0 +1,160 @@
+
+    k_rect.set_x(col - 0.5)
+    c_rect1.set_x(col + (kernel_size/2 - 1))
+    k_rect.set_y(row - 0.5)
+    c_rect1.set_y(row + (kernel_size/2 - 1))
+
+    im.set_array(all_frames[frame])
+    im.autoscale()
+
+    return [im, k_rect, c_rect1]
+
+# MAIN PROGRAM
+if __name__ == "__main__":
+
+    print("Creating blurred animation with kernel size:", kernel_size)
+
+    # Load image
+    img = plt.imread(input_file)
+
+    ### HERE WE USE THE CONVOLVE FUNCTION TO GET THE FINAL BLURRED IMAGE
+    # I chose a simple mean filter (equal kernel weights)
+    kernel = np.ones(shape=(kernel_size, kernel_size)) / kernel_size ** 2 # create kernel
+    # convolve the image i.e. apply mean filter
+    img_convolved = convolve(img, kernel, mode='constant', cval=0) # pad borders with zero like below for consistency
+
+
+    ### HERE WE CONVOLVE MANUALLY STEP-BY-STEP TO CREATE ANIMATION
+    img_pad = np.pad(img, (int(np.ceil(kernel_size/2) - 1), int(np.ceil(kernel_size/2) - 1))) # Pad image to deal with borders
+    new_img = np.zeros(img.shape, dtype=np.uint16) # this will be the blurred final image
+
+    # add first frame with complete blurred image for print version of GIF
+    all_frames = [img_convolved]
+
+    # precompute animation frames and append to the list
+    total_frames = (img_pad.shape[0] - kernel_size + 1) * (img_pad.shape[1] - kernel_size + 1) # total frames if by chance image is not squared
+    for frame in range(total_frames):
+        row = (frame % total_frames) // (img_pad.shape[0] - kernel_size + 1) # row index
+        col = (frame % total_frames) % (img_pad.shape[1] - kernel_size + 1) # col index
+        img_chunk = img_pad[row : row + kernel_size, col : col + kernel_size] # get current image chunk inside the kernel
+        new_img[row, col] = np.mean(img_chunk).astype(np.uint16) # calculate its mean -> mean filter
+        all_frames.append(new_img.copy()) # append to animation frames list
+
+    # We now have an extra frame
+    total_frames += 1 
+
+    ### FROM HERE WE START CREATING THE ANIMATION
+    # Initialize canvas
+    f, (ax1, ax2) = plt.subplots(1,2, figsize=(10,5))
+
+    # Display the padded image -> this one won't change during the animation
+    ax1.imshow(img_pad, cmap='gray')
+    # Initialize the blurred image -> this is the first frame with already the final result
+    im = ax2.imshow(img_convolved, animated=True, cmap='gray')
+
+    # Define rectangular patches to identify moving kernel
+    k_rect = p.Rectangle((-0.5,-0.5), kernel_size, kernel_size, linewidth=2, edgecolor=kernel_color, facecolor='none', alpha=0.8) # kernel rectangle
+    c_rect1 = p.Rectangle(((kernel_size/2 - 1), (kernel_size/2 - 1)), 1, 1, linewidth=2, edgecolor=center_color, facecolor='none') # central pixel rectangle
+    # Add them to the figure
+    ax1.add_patch(k_rect)
+    ax1.add_patch(c_rect1)
+
+    # Fix limits of the right image (without padding) so that it is the same size as the left image (with padding)
```

Comment:
```suggestion
    # Fix limits of the RHS image (without padding) so that it is the same size as the LHS image (with padding)
```
I find that 'right-hand side' is clearer: It's clear that it means 'as opposed to left-hand side' (whereas 'right image' could be vs. 'wrong image?').

Reply:
I see the point of the clarification, but is this a common nomenclature for referring to images? I only heard of LHS and RHS in math when talking about sides of an equation, and I would think it could be more confusing compared to natural 'left/right' natural language. But maybe I am not aware that it's already in use somewhere else in the lessons or in some other curriculums.

Reply:
@mkcor I've accepted all other suggestions. I am still unsure about this LHS/RHS change, but I can integrate it if you think it's appropriate.

Reply:
@marcodallavecchia 'RHS/LHS' are definitely technical/formal terms but they are not reserved to math equations.

I'm happy with @tobyhodges's suggestion (above) as a compromise though! :rocket:

Reply:
I'm happy with this compromise, it follows more a natural language. I will keep the same language also in the blurring episode!